In [157]:
import pandas as pd
data=pd.read_csv('C:/Users/etokarev/OneDrive - Philip Morris International/Desktop/private/Netology/Python/Homework/_23/vgsales.csv')
data.head(5)

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.53          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.76      3.79         3.29         35.52          82.0   
3     15.61     10.93      3.28         2.95         32.77          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count  User_Score  User_Count Developer Rating  
0          51.0         8.0       322.0  Nintendo      E  
1           NaN         NaN         NaN       NaN    NaN  
2          73.0         8.3       709.0  Nintendo      E  
3          73.0         8.0       192.0  Nintendo      E  
4           NaN         NaN         NaN       NaN    NaN

In [158]:
# H0- средний рейтинг критиков >= 80% (критики хорошо относятся к играм)
# H1  -средний рейтинг менее 80


from scipy import stats as st
data_upgrade=data['Critic_Score'].dropna()
print(data_upgrade.mean())

s,p=st.ttest_1samp(data_upgrade,80,alternative='less')

if p < 0.05:
    print('p= ',p,' Отвергаем H0, рейтинг менее 80%, критики плохо относятся к играм')
else:
    print ('p= ',p,' не отвергаем H0, критики хорошо относятся к играм') 

68.96767850559173
p=  0.0  Отвергаем H0, рейтинг менее 80%, критики плохо относятся к играм


In [159]:
#HO -критикам все равно, им одинаково нравятся стрелялки и стратегии
#H1 -критикам нравятся больше нравиятся стратегии

data_pc=data[data['Genre'] == "Shooter"]['Critic_Score'].dropna()
data_ps4=data[data['Genre']== "Strategy"]['Critic_Score'].dropna()
s,p= st.ttest_ind(data_pc, data_ps4)
if p< 0.05:
    print('p= ',p, ' при уровне значимости 5% H0 отвергаем, критикам больше нравятся стратегии')
else: 
    print ('p= ',p,' при уровне значимости 5% H0 не отвергаем,  критикам одинаково нравятся стрелялки и стратегии')

p=  0.03896871303135602  при уровне значимости 5% H0 отвергаем, критикам больше нравятся стратегии


In [160]:
data[data['Platform'].isin(['PC','PS4'])].groupby(data['Platform']).mean()['Critic_Score']

Platform
PC     75.928671
PS4    72.091270
Name: Critic_Score, dtype: float64

In [161]:
#HO -критикам все равно, им одинаково нравятся PC  и PS4
#H1 -нравятся больше игры на PC 

data_pc=data[data['Platform']=="PC"]['Critic_Score'].dropna()
data_ps4=data[data['Platform']=="PS4"]['Critic_Score'].dropna()
s,p= st.ttest_ind(data_pc, data_ps4)
if p< 0.05:
    print('p= ',p, ' при уровне значимости 5% H0 отвергаем,им больше нравятся PC')
else:
    print ('p= ',p, ' при уровне значимости 5% H0 не отвергаем,им одинаково нравятся PC и PS4')

p=  6.215797831285281e-06  при уровне значимости 5% H0 отвергаем,им больше нравятся PC


In [162]:
spam_data=pd.read_csv('C:/Users/etokarev/OneDrive - Philip Morris International/Desktop/private/Netology/Python/Homework/_23/spam.csv')
spam_data.head(5)

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [163]:
#1) Привидите весь текст к нижнему регистру
def data_lower(row):
    return row['Message'].lower()

spam_data['Message']=spam_data.apply(data_lower, axis=1)
spam_data.head(5)

Category                                            Message
0      ham  go until jurong point, crazy.. available only ...
1      ham                      ok lar... joking wif u oni...
2     spam  free entry in 2 a wkly comp to win fa cup fina...
3      ham  u dun say so early hor... u c already then say...
4      ham  nah i don't think he goes to usf, he lives aro...

In [164]:
#2) Удалите мусорные символы;
import re
def data_only_words(row):
    return re.sub('[\W_]+',' ', row['Message'])

spam_data['Message']=spam_data.apply(data_only_words, axis=1)
spam_data.head(5)

Category                                            Message
0      ham  go until jurong point crazy available only in ...
1      ham                           ok lar joking wif u oni 
2     spam  free entry in 2 a wkly comp to win fa cup fina...
3      ham       u dun say so early hor u c already then say 
4      ham  nah i don t think he goes to usf he lives arou...

In [165]:
#conda install nltk.corpus
#import nltk
#nltk.download('stopwords')

In [166]:
#3) Удалите стоп-слова;
from nltk.corpus import stopwords
stopwords_set = set(stopwords.words('english'))

def data_wo_stopwords(row):
    new_text=''
    spitted_text = row['Message'].split()
    for word in spitted_text:
        if word not in stopwords_set:
            new_text =new_text + ' '+ word
    return new_text

spam_data['Message']=spam_data.apply(data_wo_stopwords, axis=1)
spam_data.head(5)


Category                                            Message
0      ham   go jurong point crazy available bugis n great...
1      ham                            ok lar joking wif u oni
2     spam   free entry 2 wkly comp win fa cup final tkts ...
3      ham                u dun say early hor u c already say
4      ham             nah think goes usf lives around though

In [167]:
#nltk.download('wordnet')

In [168]:
#4) Привидите все слова к нормальной форме;
from nltk.stem import WordNetLemmatizer
wordnet_lemmatizer = WordNetLemmatizer()

def lemma(row):
    new_text=''
    spitted_text = row['Message'].split()
    for word in spitted_text:
        new_text = new_text+' '+ wordnet_lemmatizer.lemmatize(word)
    return new_text    
        
spam_data['Message']=spam_data.apply(lemma, axis=1)  
spam_data.head(5)

Category                                            Message
0      ham   go jurong point crazy available bugis n great...
1      ham                            ok lar joking wif u oni
2     spam   free entry 2 wkly comp win fa cup final tkts ...
3      ham                u dun say early hor u c already say
4      ham                nah think go usf life around though

In [169]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(spam_data['Message'])
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=names)
tfidf_matrix

00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
1     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
2     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
3     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
4     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
...   ...  ...     ...           ...   ...   ...          ...          ...   
5567  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5568  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5569  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5570  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5571  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   

      0125698789   02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
1            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
2            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
3            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
4            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
...          ...  ...  ...    ...     ...  ...        ...   ...   ...    ...   
5567         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5568         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5569         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5570         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5571         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   

       èn   ú1  〨ud  
0     0.0  0.0  0.0  
1     0.0  0.0  0.0  
2     0.0  0.0  0.0  
3     0.0  0.0  0.0  
4     0.0  0.0  0.0  
...   ...  ...  ...  
5567  0.0  0.0  0.0  
5568  0.0  0.0  0.0  
5569  0.0  0.0  0.0  
5570  0.0  0.0  0.0  
5571  0.0  0.0  0.0  

[5572 rows x 8048 columns]

In [170]:
from  sklearn.model_selection import train_test_split


x = tfidf_matrix
y = spam_data['Category']

x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.3,random_state=42)

In [171]:
from sklearn.linear_model import LogisticRegression

model=LogisticRegression(random_state=42)
model.fit(x_train, y_train)


LogisticRegression(random_state=42)

In [172]:
from sklearn.metrics import accuracy_score
accuracy_score (y_test, model.predict(x_test))

0.958732057416268

In [173]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, model.predict(x_test),labels=["spam", "ham"])

array([[ 158,   66],
       [   3, 1445]], dtype=int64)

158-это на самом деле спам и модель показала, что это спам; 1445 - это не спам,  и модель показала, что это не спам
66 -это не спам,  а модель  показала, что спам; 3 спам, а модель показала, что не спам

In [174]:
spam_data['Categoty_Predict']=model.predict(tfidf_matrix)
spam_data

Category                                            Message  \
0         ham   go jurong point crazy available bugis n great...   
1         ham                            ok lar joking wif u oni   
2        spam   free entry 2 wkly comp win fa cup final tkts ...   
3         ham                u dun say early hor u c already say   
4         ham                nah think go usf life around though   
...       ...                                                ...   
5567     spam   2nd time tried 2 contact u u 750 pound prize ...   
5568      ham                        ü b going esplanade fr home   
5569      ham                               pity mood suggestion   
5570      ham   guy bitching acted like interested buying som...   
5571      ham                                     rofl true name   

     Categoty_Predict  
0                 ham  
1                 ham  
2                spam  
3                 ham  
4                 ham  
...               ...  
5567             spam  
5568              ham  
5569              ham  
5570              ham  
5571              ham  

[5572 rows x 3 columns]

In [187]:
def incorrect_model(row):
    if row['Category'] != row['Categoty_Predict']:
        dict_spam.append([row['Category'],row['Categoty_Predict'], row['Message']])

dict_spam=[]            
spam_data.apply(incorrect_model, axis=1)
result = pd.DataFrame(dict_spam, columns=['Category', 'Category_Predict', 'Message'])
result

Category Category_Predict  \
0       spam              ham   
1       spam              ham   
2       spam              ham   
3       spam              ham   
4       spam              ham   
..       ...              ...   
200     spam              ham   
201     spam              ham   
202     spam              ham   
203     spam              ham   
204     spam              ham   

                                               Message  
0     freemsg hey darling 3 week word back like fun...  
1     xxxmobilemovieclub use credit click wap link ...  
2     england v macedonia dont miss goal team news ...  
3     sm ac sptv new jersey devil detroit red wing ...  
4     congrats 1 year special cinema pas 2 call 090...  
..                                                 ...  
200   latest news police station toilet stolen cop ...  
201   sparkling shopping break 45 per person call 0...  
202   http tm widelive com index wml id 820554ad0a1...  
203   want explicit sex 30 sec ring 02073162414 cos...  
204   asked 3mobile 0870 chatlines inclu free min i...  

[205 rows x 3 columns]